CancelablePromise.js

```diff
@@ -19,19 +19,36 @@ export default class CancelablePromise {
   }
 
   then(success, error) {
-    return new CancelablePromise((resolve, reject) => {
+    const p = new CancelablePromise((resolve, reject) => {
       this._promise.then((r) => {
+        if (this._canceled) {
+          p.cancel();
+          resolve(r);
+        }
         if (success && !this._canceled) {
+          let argument = r;
           const returned = success(r);
-          resolve(returned || r);
+          if (typeof returned !== 'undefined') {
+            argument = returned;
+          }
+          resolve(argument);
```

Comment:
if returned is `undefined`, we should call `resolve()`, not forward the current argumet. What do you think?